Add create_type helper function

jkimbo opened the pull request:
<!--- Provide a general summary of your changes in the title above. -->

<!--- This template is entirely optional and can be removed, but is here to help both you and us. -->
<!--- Anything on lines wrapped in comments like these will not show up in the final text. -->

## Description

Add a new `create_type` function in a new `strawberry.tools` module.

## Types of Changes

<!--- What types of changes does your pull request introduce? Put an `x` in all the boxes that apply. -->
- [ ] Core
- [ ] Bugfix
- [x] New feature
- [ ] Enhancement/optimization
- [ ] Documentation

## Issues Fixed or Closed by This PR

*

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
- [x] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

botberry commented:
Hi, thanks for contributing to Strawberry 🍓!

We noticed that this PR is missing a `RELEASE.md` file. We use that to automatically do releases here on GitHub and, most importantly, to PyPI!

So as soon as this PR is merged, a release will be made 🚀.

Here's an example of `RELEASE.md`:

```markdown
Release type: patch

Description of the changes, ideally with some examples, if adding a new feature.
```

Release type can be one of patch, minor or major. We use [semver](https://semver.org/),so make sure to pick the appropriate type. If in doubt feel free to ask :)
<!-- action-check: release-file -->

botberry commented:
Hi, thanks for contributing to Strawberry 🍓!

We noticed that this PR is missing a `RELEASE.md` file. We use that to automatically do releases here on GitHub and, most importantly, to PyPI!

So as soon as this PR is merged, a release will be made 🚀.

Here's an example of `RELEASE.md`:

```markdown
Release type: patch

Description of the changes, ideally with some examples, if adding a new feature.
```

Release type can be one of patch, minor or major. We use [semver](https://semver.org/),so make sure to pick the appropriate type. If in doubt feel free to ask :)
<!-- action-check: release-file -->

botberry commented:
Hi, thanks for contributing to Strawberry 🍓!

We noticed that this PR is missing a `RELEASE.md` file. We use that to automatically do releases here on GitHub and, most importantly, to PyPI!

So as soon as this PR is merged, a release will be made 🚀.

Here's an example of `RELEASE.md`:

```markdown
Release type: patch

Description of the changes, ideally with some examples, if adding a new feature.
```

Release type can be one of patch, minor or major. We use [semver](https://semver.org/),so make sure to pick the appropriate type. If in doubt feel free to ask :)
<!-- action-check: release-file -->

BryceBeagle reviewed (approved):
Looks good to me. But, as I mentioned on Discord, hopefully this becomes unnecessary with the implementation of `StrawberryObjectType` :+1:

patrick91 commented on docs/guides/tools.md:
```diff
@@ -0,0 +1,54 @@
+
+@strawberry.field
+def get_user_by_username(username: str) -> User:
+    user = # get user
```
```suggestion
    user = ...  # get user
```

patrick91 commented on docs/guides/tools.md:
```diff
@@ -0,0 +1,54 @@
+
+@strawberry.mutation
+def create_user(username: str) -> User:
+    user = # create user
```
```suggestion
    user = ...  # create user
```

patrick91 commented on strawberry/tools/create_type.py:
```diff
@@ -0,0 +1,21 @@
+from strawberry.field import StrawberryField
+
+
+def create_type(name: str, fields: List[StrawberryField]) -> Type:
```
let's add a docstring 😊

do you think we should add some safety checks here? I'm thinking of checking at runtime if fields are actual fields and also preventing mixing fields with mutations, what do you think?

jkimbo replied:
Runtime checks sound like a good idea 👍

marcoacierno replied:
Also maybe a check to make sure `fields` is not empty? (at least when creating a Query 🤔 )

jkimbo replied:
@marcoacierno we can't tell if you're creating a Query (since it's the same as any other type) but I think it makes sense to always raise an exception if the function is passed an empty list of fields.

jkimbo replied:
@patrick91 I like the idea of preventing fields and mutations but AFAIK currently we can't tell if a field is a mutation or now: https://github.com/strawberry-graphql/strawberry/blob/f31e6c17238ba4372129c4f291ddf336b4e6b901/strawberry/mutation.py#L8

marcoacierno commented on tests/tools/test_create_type.py:
```diff
@@ -0,0 +1,75 @@
+    class User:
+        username: str
+
+    @strawberry.mutation
```
I would add another test that checks the parameters (e.g permission_classes=[X]) passed to  `@strawberry.mutation()` are not lost

(this is coming from experience where in a previous version of the create_type function I had this bug :D)

jkimbo replied:
Good idea!

marcoacierno reviewed (commented):
Nice work, the only thing I am worried about is that once we add this way we will have to support it as some people will use it, and if in the future we add the schema builder system, we will have 2 different ways to do the same thing

But I guess it's something we can worry about later 😄

jkimbo commented:
@marcoacierno @BryceBeagle @patrick91 this is ready for a review again. I've added the runtime checks and rewritten the example in the docs so that it's less complicated (and more focussed on the function).

patrick91 commented on RELEASE.md:
```diff
@@ -0,0 +1,21 @@
+Release type: minor
+
+Add function `create_type` to create a Strawberry type from a list of fields.
```
```suggestion
This release adds a function called `create_type` to create a Strawberry type from a list of fields.
```

patrick91 commented on docs/guides/tools.md:
```diff
@@ -0,0 +1,44 @@
+
+### `create_type`
+
+Create a Strawberry type from a list of StrawberryFields.
```
```suggestion
Create a Strawberry type from a list of fields.
```

what do you think of renaming `StrawberryFields` to fields here? I think someone that's new to the library might not know what a `StrawberryField` is

BryceBeagle replied:
Staying agnostic to the implementation here seems reasonable to me

jkimbo replied:
Yep makes sense.

patrick91 reviewed (approved):
Looks good to me! I've added a couple of minor notes, but we can merge anytime :)
